config/src/main/java/com/alibaba/nacos/config/server/service/kubernetes/ConfigMapSyncServiceImpl.java

```diff
@@ -0,0 +1,352 @@
+        startWatchConfigMap();
+        Runtime.getRuntime().addShutdownHook(new Thread(() -> {
+            Loggers.MAIN.info("[{}] stopping...", "configMap-sync");
+            ConfigMapSyncServiceImpl.this.stop();
+            Loggers.MAIN.info("[{}] stopped.", "configMap-sync");
+        }));
+    }
+    
+    /**
+     * use the Java API from an application outside a kubernetes cluster.
+     * you should load a kubeConfig to generate apiClient instead of getting it from coreV1api.
+     */
+    public ApiClient getOutsideApiClient() throws IOException {
+        String kubeConfigPath = configMapSyncConfig.getKubeConfig();
+        
+        // loading the out-of-cluster config, a kubeconfig from file-system
+        ApiClient apiClient = ClientBuilder.kubeconfig(KubeConfig.loadKubeConfig(new FileReader(kubeConfigPath))).build();
+        
+        // set the global default api-client to the in-cluster one from above
+        Configuration.setDefaultApiClient(apiClient);
+        return apiClient;
+    }
+    
+    /**
+     *  Start watch ConfigMap.
+     */
+    public void startWatchConfigMap() throws IOException {
+        ApiClient apiClient;
+        CoreV1Api coreV1Api;
+        if (configMapSyncConfig.isOutsideCluster()) {
+            Loggers.MAIN.info("[{}] use outside cluster Apiclient.", "configMap-sync");
+            apiClient = getOutsideApiClient();
+            coreV1Api = new CoreV1Api(apiClient);
+        } else {
+            Loggers.MAIN.info("[{}] use local cluster Apiclient.", "configMap-sync");
+            coreV1Api = new CoreV1Api();
+            apiClient = coreV1Api.getApiClient();
+        }
+        OkHttpClient httpClient = apiClient.getHttpClient().newBuilder().readTimeout(Duration.ZERO).build();
+        apiClient.setHttpClient(httpClient);
+        factory = new SharedInformerFactory(apiClient);
+        SharedIndexInformer<V1ConfigMap> configInformer = factory.sharedIndexInformerFor(
+                (CallGeneratorParams params) -> {
+                    return coreV1Api.listConfigMapForAllNamespacesCall(true, null, null, null, null, null,
+                            params.resourceVersion, null, params.timeoutSeconds, params.watch, null);
+                }, V1ConfigMap.class, V1ConfigMapList.class);
+        configInformer.addEventHandler(new ResourceEventHandler<V1ConfigMap>() {
+            @Override
+            public void onAdd(V1ConfigMap obj) {
+                if (obj == null || obj.getMetadata() == null || obj.getData() == null) {
+                    return;
+                }
+                Loggers.MAIN.info("[{}] adding configMap...", "configMap-sync");
+                ConfigInfo configInfo = configMapToNacosConfigInfo(obj);
+                String srcIp = apiClient.getBasePath();
+                
+                try {
+                    publishConfigMap(obj, srcIp);
+                } catch (NacosException e) {
+                    throw new RuntimeException(e);
```

Comment:
Here I prefer error logging rather than throwing exceptions directly, considering the robustness of thread running.

----

这里我倾向于error日志而不是直接throw异常，考虑线程运行的健壮性。